lark/lark.py

```diff
@@ -23,41 +19,51 @@
 except ImportError:
     regex = None
 
+
+###{standalone
+from typing import (
+    TypeVar, Type, List, Dict, Iterator, Callable, Union, Optional,
+    Tuple, Iterable, IO, Any, TYPE_CHECKING
+)
+
 if TYPE_CHECKING:
-    from .load_grammar import PackageResource
-    from .exceptions import UnexpectedInput
     from .parsers.lalr_interactive_parser import InteractiveParser
     from .visitors import Transformer
-
-###{standalone
+    if sys.version_info >= (3, 8):
+        from typing import Literal
+    else:
+        from typing_extensions import Literal
 
 class PostLex(ABC):
     @abstractmethod
-    def process(self, stream):
+    def process(self, stream: Iterator[Token]) -> Iterator[Token]:
         return stream
 
-    always_accept = ()
+    always_accept: Iterable[str] = ()
 
 class LarkOptions(Serialize):
     """Specifies the options for Lark
 
     """
 
     start: List[str]
-    parser: str
-    lexer: str
-    transformer: 'Optional[Transformer]'
-    postlex: Optional[PostLex]
-    ambiguity: str
-    regex: bool
     debug: bool
-    keep_all_tokens: bool
+    transformer: 'Optional[Transformer]'
     propagate_positions: Union[bool, str]
     maybe_placeholders: bool
-    lexer_callbacks: Dict[str, Callable[[Token], Token]]
     cache: Union[bool, str]
+    regex: bool
     g_regex_flags: int
+    keep_all_tokens: bool
+    tree_class: Any
+    parser: 'Literal["earley", "lalr", "cyk", "auto"]'
+    lexer: 'Union[Literal["auto", "standard", "contextual", "dynamic", "dynamic_complete"], Type[Lexer]]'
```

Comment:
We can instead define types inside a version guarded if-statement that when `Literal` is not available is just `Union[str, Type[Lexer]]` (and similar for all other types here). Not sure if it is better thou.

Reply:
I prefer the current way. It provides the precision of `Literal` even when type checking on 3.6 and 3.7.

Reply:
What does the current implementation do when `typing_extensions` is not installed?

Reply:
The current implementation executes normally because the import of `typing_extensions` is wrapped in an `if TYPE_CHECKING`. I believe it is reasonable to assume that `typing_extensions` will be installed when `TYPE_CHECKING` is true because `typing_extensions` is a dependency of `mypy`.